tools/codegen/api/python.py

```diff
@@ -0,0 +1,987 @@
+# cases, e.g.:
+#
+#   // aten::add.Tensor(Tensor self, Tensor other, *, Scalar alpha=1) -> Tensor
+#   [](const Tensor & self, const Tensor & other, Scalar alpha) -> Tensor
+#
+# For out variant the 'out' argument's type is changed from 'Tensor &'
+# to 'Tensor'. It's because when calling the lambda it passes in the
+# PythonArgParser output '_r.tensor(3)', which is stack allocated object
+# and needs to pass by value.
+#
+#   // aten::add.out(Tensor self, Tensor other, *, Scalar alpha=1, Tensor(a!) out) -> Tensor(a!)
+#   [](Tensor out, const Tensor & self, const Tensor & other, Scalar alpha) -> Tensor
+#
+# For multi-output case it can keep using reference type because the
+# PythonArgParser output has been unpacked to local variables, e.g.:
+#
+#   // aten::max.names_dim_max(Tensor self, Dimname dim, bool keepdim=False, *,
+#   //     Tensor(a!) max, Tensor(b!) max_values) -> (Tensor(a!) values, Tensor(b!) indices)
+#   [](Tensor & max, Tensor & max_values, const Tensor & self, Dimname dim, bool keepdim) -> std::tuple<Tensor,Tensor>
+#
+# For deprecated python signature, it should follow deprecated python arg order.
+# TODO: This is to keep same byte-for-byte result as the old codegen - maybe unnecessary?
+
+def dispatch_lambda_args(f: NativeFunction, method: bool, *,
+                         python_signature: Optional[PythonSignature] = None,
+                         ) -> Tuple[DispatchLambdaArgument, ...]:
+    # Start with cpp arguments - dispatch lambda signature always include 'self'
+    cpp_args: Sequence[CppArgument] = _cpp_signature(f, method=False).arguments()
+
+    # Special reorder logic for deprecated python signature
+    if isinstance(python_signature, PythonSignatureDeprecated):
+        m: Dict[str, CppArgument] = dict((a.name, a) for a in cpp_args)
+        # reorder according to the deprecated signature
+        # ignore 'out' argument when binding to non-output function.
+        ordered_args = filter(lambda n: n != 'out' or f.func.is_out_fn(),
+                              python_signature.deprecated_args_names)
+        cpp_args = list(map(lambda n: m[n], ordered_args))
+
+    out_args: Set[str] = set(a.name for a in f.func.out_arguments)
+
+    # Convert from cpp argument to lambda argument
+    def dispatch_lambda_arg(cpp_arg: CppArgument) -> DispatchLambdaArgument:
+        type_str = cpp_arg.type
+        is_out_arg = cpp_arg.name in out_args
+        if method and cpp_arg.name == 'self':
+            # For method's 'self', we can use 'Tensor &' and simply ignore mutability!
+            type_str = 'Tensor &'
+        else:
+            # For other cases we need prevent dangling refs to temps (unless it's
+            # unpacked scattered output)
+            ensure_temp_safe = len(out_args) <= 1 or not is_out_arg
+            if ensure_temp_safe:
+                type_str = {
+                    'Tensor &': 'Tensor',
+                }.get(type_str, type_str)
```

Comment:
This is the thing which makes us gen `Tensor out` and not `Tensor& out` at

```
    // aten::_bmm.out(Tensor self, Tensor mat2, *, bool deterministic=False, Tensor(a!) out) -> Tensor(a!)
    auto dispatch__bmm_out = [](Tensor out, const Tensor & self, const Tensor & mat2, bool deterministic) -> Tensor {
      pybind11::gil_scoped_release no_gil;
      return at::_bmm_out(out, self, mat2, deterministic);
    };
    return wrap(dispatch__bmm_out(_r.tensor(3), _r.tensor(0), _r.tensor(1), _r.toBool(2)));
```

Which is indirectly explained by the bottom of the comment at "Use the simple_return_type (Tensor) rather than the fancy return type".  We should probably have a reference to that comment. We should also restructure the codegen structure here so that this special case isn't necessary at all. (Not for this PR, because that wouldn't be byte-for-byte compatible)

Reply:
> This is the thing which makes us gen `Tensor out` and not `Tensor& out` at

In the comment above this method I also called out this special case:
```
# For out variant the 'out' argument's type is changed from 'Tensor &'
# to 'Tensor'. It's because when calling the lambda it passes in the
# PythonArgParser output '_r.tensor(3)', which is stack allocated object
# and needs to pass by value.
#
#   // aten::add.out(Tensor self, Tensor other, *, Scalar alpha=1, Tensor(a!) out) -> Tensor(a!)
#   [](Tensor out, const Tensor & self, const Tensor & other, Scalar alpha) -> Tensor
```

I'll add some inline comment here to make it more prominent.

Reply:
inline comments added